File: Hashing.py
class UserHashTable:
    def __init__(self, size=10):
        """Initialize the hash table with a given size."""
        self.size = size
        self.table = [[] for _ in range(size)]

    def _hash_function(self, username):
        """Hash function to map a username to an index."""
        return hash(username) % self.size

    def add_user(self, username, profile_data):
        """Add a user with the given username and profile data."""
        index = self._hash_function(username)
        # Check if the username already exists in the chain and update it if so
        for i, (existing_username, _) in enumerate(self.table[index]):
            if existing_username == username:
                self.table[index][i] = (username, profile_data)
                print(f"Updated user {username} at index {index}")
                return
        # Append new user if not found
        self.table[index].append((username, profile_data))
        print(f"Added user {username} at index {index}")

    def get_user(self, username):
        """Retrieve a user's profile data by username."""
        index = self._hash_function(username)
        for existing_username, profile_data in self.table[index]:
            if existing_username == username:
                return profile_data
        print(f"User {username} not found.")
        return None

# Example usage
user_table = UserHashTable()
user_table.add_user("alice123", {"name": "Alice", "age": 25, "location": "NY"})
user_table.add_user("bob456", {"name": "Bob", "age": 30, "location": "LA"})
print(user_table.get_user("alice123"))  # Should print Alice's profile
print(user_table.get_user("charlie789"))  # Should print "User charlie789 not found."

File: Hashing.cpp
#include <iostream>
#include <vector>
#include <list>
#include <string>
#include <utility>

using namespace std;

class UserHashTable {
private:
    int size;
    vector<list<pair<string, string>>> table;

    int hashFunction(const string& username) const {
        // Simple hash function using std::hash and modulo operation
        return hash<string>()(username) % size;
    }

public:
    UserHashTable(int tableSize = 10) : size(tableSize) {
        table.resize(size);
    }

    void addUser(const string& username, const string& profileData) {
        int index = hashFunction(username);
        // Check if the username already exists and update if so
        for (auto& pair : table[index]) {
            if (pair.first == username) {
                pair.second = profileData;
                cout << "Updated user " << username << " at index " << index << endl;
                return;
            }
        }
        // Add new user if not found
        table[index].emplace_back(username, profileData);
        cout << "Added user " << username << " at index " << index << endl;
    }

    string getUser(const string& username) const {
        int index = hashFunction(username);
        for (const auto& pair : table[index]) {
            if (pair.first == username) {
                return pair.second;
            }
        }
        cout << "User " << username << " not found." << endl;
        return "";
    }
};

// Example usage
int main() {
    UserHashTable userTable;
    userTable.addUser("alice123", "Alice, 25, NY");
    userTable.addUser("bob456", "Bob, 30, LA");
    cout << userTable.getUser("alice123") << endl; // Should print Alice's profile data
    cout << userTable.getUser("charlie789") << endl; // Should print "User charlie789 not found."
    return 0;
}
